### PR TITLE
Enhancement LEfSe

### PR DESCRIPTION
**Pull request by camillegoldman**
# Pull Request Template for MMEDS

## What has changed
Added a lefse.py file to create a script to run lefse analysis
Updated server.py and html files to support lefse uploads

## Checklist of pre-requisites
-   [x] Does the code run?  
-   [ ] Does the code follow the repository style?  
-   [x] Is the code tested?  

## How to use the feature
Example: To do X run new function Y with parameters A, B, and C which gives result Z.

## How the design has changed
Example: Functions A, B, and C are now all methods of Class D.

### CODE REVIEW

File: mmeds/lefse.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        load = 'module use {}/.modules/modulefiles; module load lefse;'.format(DATABASE_DIR.parent)
+        self.jobtext.append(load)
+        self.jobtext.append('{}={};'.format(str(self.run_dir).replace('$', ''), self.path))
+        self.module = load
+
+        self.subclass = 'subclass' in self.doc.reads_type
+        self.subjects = 'subjects' in self.doc.reads_type
+
+    def format_input(self):
+        """Convert uploaded .txt file into file type usable by LEfSe"""
```
Comment (DSWallach) [inline mmeds/lefse.py]: Spaces around the docstrings 
```suggestion
        """ Convert uploaded .txt file into file type usable by LEfSe """
```

File: mmeds/server.py [hunk 2]
```diff
@@ -553,13 +602,19 @@ def __init__(self, watcher, q, testing=False):
...
         ----------------------------------------
         :access_code: The code that identifies the dataset to run the tool on
         :tool: The tool to run on the chosen dataset
         """
+        if '-' in tool:
+            tool_type = tool.split('-')[0]
+            analysis_type = tool.split('-')[1]
+        else:
+            tool_type = tool
+            analysis_type = 'default'
```
Comment (DSWallach) [inline mmeds/server.py]: I am moving away from the combined properties for tool, so `tool_type` and `analysis_type` should be separate parameters

File: mmeds/tests/unit/test.py [hunk 3]
```diff
@@ -102,6 +102,15 @@ def setup_tests(tests):
                         fig.TEST_OTU,
                         fig.TEST_CODE_OTU)
             assert 0 == upload_otu(test_otu)
+            #Upload Lefse data if running test_tools.py
```
Comment (DSWallach) [inline mmeds/tests/unit/test.py]: Spaces before comments
```suggestion
            # Upload Lefse data if running test_tools.py
```

File: mmeds/tests/unit/test_spawn.py [hunk 4]
```diff
@@ -81,7 +81,7 @@ def test_b_start_analysis(self):
         # Check the process exited with code 0
         self.assertEqual(self.pipe.recv(), 0)
         self.assertEqual(self.pipe.recv(), 0)
-
+
```
Comment (DSWallach) [inline mmeds/tests/unit/test_spawn.py]: remove whitespace

File: mmeds/tests/unit/test_tools.py [hunk 5]
```diff
@@ -55,10 +60,18 @@ def test_qiime2_child_setup_analysis(self):
...
+        q2.create_analysis(Lefse)
+        rmtree(q2.path)
+    
     def test_sparcc_sub_analysis(self):
         config = load_config(Path(fig.TEST_CONFIG).read_text(), fig.TEST_METADATA)
         q2 = Qiime2(fig.TEST_USER, fig.TEST_CODE_SHORT, 'qiime2', 'dada2', config, testing=self.testing, analysis=False)
         q2.setup_analysis()
         q2.create_analysis(SparCC)
         rmtree(q2.path)
+
```
Comment (DSWallach) [inline mmeds/tests/unit/test_tools.py]: remove whitespace

File: mmeds/tests/unit/test_tools.py [hunk 6]
```diff
@@ -29,6 +30,10 @@ def run_qiime(self, code, tool_type, analysis_type, data_type, Qiime):
     def test_sparcc_setup_analysis(self):
         self.run_qiime(fig.TEST_CODE_OTU, 'sparcc', 'default', 'otu_table', SparCC)
```
Comment (DSWallach) [inline mmeds/tests/unit/test_tools.py]: One newline before a method in a class. Two newlines before a function

File: mmeds/server.py [hunk 7]
```diff
@@ -669,6 +672,17 @@ def process_data(self, public=False, **kwargs):
             datafiles = self.load_data_files(otu_table=kwargs['otu_table'])
             reads_type = None
             barcodes_type = None
+        elif cp.session['upload_type'] == 'lefse':
+            datafiles = self.load_data_files(lefse_table=kwargs['lefse_table'])
+            if 'subclass' in kwargs.items():
+                reads_type = 'subclass'
+                if 'subjects' in kwargs.items():
+                    reads_type = reads_type + '_subjects'
+            elif 'subjects' in kwargs.items():
```
Comment (DSWallach) [inline mmeds/server.py]: I think this should be `x in kwargs.keys()`. `.items()` returns `(key, value)` pairs so `in` would try to match against that

File: mmeds/server.py (hunk 2 above)
Comment (camillegoldman) [reply mmeds/server.py]: 'tool' is the input received from the html file 'analysis_select_tool.html' There is only one drop down on that page, and the response right now is in the format of tool_type-analysis_type (or just tool_type for sparcc and lefse). This bit of code split up the input into the respective variables, so 'tool' is never used after this. 
If it's really important that tool_type and analysis_type be defined from the input directly, I can change the html page to have two drop down menus, one for tool_type and one for analysis_type. However, I thought that the one drop down menu and then splitting the input into two in server.py is more user friendly. I can rename the variable something other than tool, to make it more clear.

Comment (DSWallach) [reply mmeds/server.py]: I think this is okay for now. Looks like I didn't update the html in the last branch I merged to master. Eventually I'd like to have one drop down be the tool and then based on which tool was selected there would pop up a second menu for additional options which would then be written to the config file for that analysis run. I think that's outside the scope of this PR though

**Pull request merged**